xayn-ai/src/ranker/context.rs

```diff
@@ -0,0 +1,117 @@
+use crate::{data::document_data::CoiComponent, ranker::utils::Id};
+
+/// The context used to calculate a document's score.
+/// https://xainag.atlassian.net/wiki/spaces/M2D/pages/770670607/Production+AI+Workflow#3.2-Context-calculations.
+/// outlines the calculation of positive and negative distance factor.
+pub(crate) struct Context {
+    /// Average positive distance.
+    pos_avg: f32,
+    /// Maximum negative distance.
+    neg_max: f32,
+}
+
+impl Context {
+    pub(crate) fn from_cois(cois: &[(Id, CoiComponent)]) -> Self {
+        let cois_len = cois.len() as f32;
+        let pos_avg = cois.iter().map(|(_, coi)| coi.pos_distance).sum::<f32>() / cois_len;
```

Comment:
if cois is empty, then the pos_avg will be nan, is that ok for the context or should it be handled?

Reply:
I guess it would be fine. cois can only be empty when `documents` is empty and when we have no documents we don't call `calculate_score`. We can add a check at the the beginning of `rank` and return early if the `documents` is empty wdyt?